homeworks/solutions/krasnor/5-remesh/src/halfedge.js

```diff
@@ -0,0 +1,821 @@
+    }
+
+    // ## 5. doEdgeCollapse
+    let cnt_total_pairs = validPairs.length;
+    let cnt_total_edges = this.edges.length;
+    let cnt_total_faces = this.faces.length;
+    let cnt_total_vertices = this.vertices.length;
+    let reduce_face_goal = cnt_total_faces-Math.round(cnt_total_faces*reduceRatio); // ratio 0 = nothing to reduce, ratio 1 = all to reduce
+    let reduce_face_goal_unrounded = cnt_total_faces-(cnt_total_faces*reduceRatio);
+
+    let curr_face_cnt = cnt_total_faces;
+
+    let nxt_idx_vertices = this.vertices.length;
+    let nxt_idx_halfedges = this.halfedges.length;
+    let nxt_idx_edges = this.edges.length;
+    let nxt_idx_faces = this.faces.length;
+
+    // ## Statistics
+    console.log("============== Statistics ======================")
+    console.log("Vertices: %d, Edges: %d, Faces: %d", cnt_total_vertices, cnt_total_edges, cnt_total_faces);
+    console.log("reduce goal: %s, reduce_unrounded: %s", reduce_face_goal, reduce_face_goal_unrounded);
+    // console.log("Reducing from %s -> %s", this.vertices.length, "not set");
+    console.log("Error Value Range: [%s,%s]", dbg_min_error, dbg_max_error);
+    console.log("Qbar invertable: %s", dbg_cnt_mat_invertable);
+    console.log("Qbar NOT invertable: %s", dbg_cnt_mat_not_invertable);
+
+    console.log("========= Calculated Errors (initial before decimation) ========")
+    let dbg_cnt_collapse_entries_after = 8;
+    let dbg_need_to_collapse = dbg_cnt_collapse_entries_after*2 <= cnt_total_pairs;
+    for(let i = 0; i < cnt_total_pairs; i++){
+      let p = dbg_heap.pop();
+      if(i < dbg_cnt_collapse_entries_after || i > (cnt_total_pairs-dbg_cnt_collapse_entries_after-1)){
+        console.log("%d - %s: %s", i, p.dbg_Qbar_Invertible, p.error);
+      }
+      if(i == dbg_cnt_collapse_entries_after && dbg_need_to_collapse){
+        console.log("...")
+      }
+    }
+    console.log("========= Starting to Collapse - current face count: %s - goal: %s", curr_face_cnt, reduce_face_goal);
+
+    for(let i = 0; (curr_face_cnt > reduce_face_goal) && (curr_face_cnt > 0); i++){
+      let faces_removed = 0;
+      let p = heap.pop();
+      let e = p.pair; // Edge
+      let m = p.mergePoint;
+      if(e.idx === -1)
+        continue; // outdated edge, ignore
+
+      let real_face_cnt = 0;
+      let deleted_face_cnt = 0;
+      this.faces.forEach((v_f, index) => {
+          if(v_f.idx != -1){
+            real_face_cnt++;
+          }else{
+            deleted_face_cnt++;
+          }
+        }
+      );
+
+      console.log("in collapse: %s - current cnt faces %s || real face count: %s, deleted: %s", i, curr_face_cnt, real_face_cnt, deleted_face_cnt);
+      let curr_hf = e.halfedge;
+      let curr_hf_twin = curr_hf.twin;
+      let curr_p1 = curr_hf.vertex;
+      let curr_p2 = curr_hf_twin.vertex;
+
+      // create new vertex along edge
+      let new_vertex = new Vertex();
+      new_vertex.idx = nxt_idx_vertices++;
```

Comment:
You could simply use the `this.vertices.length` as the next index of a vertex :)